src/Corax/Utils/EntryIdEncodings.cs

```diff
@@ -0,0 +1,112 @@
+using System;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+namespace Corax.Utils;
+
+public static class EntryIdEncodings
+{
+    //Schema:
+    //[CONTAINER_ID][FREQUENCY][CONTAINER_TYPE]
+    
+    // Positions:
+    private const int FrequencySizeInBits = 8;
+    private const int EntryIdOffset = FrequencySizeInBits + ContainerTypeOffset;
+    private const long Mask = 0xFFL;
+    private const int ContainerTypeOffset = 2;
+    
+    // Quantization parameters:
+    private const long Min = 0;
+    private const long Max = 255;
+    private const long QuantizationMax = 256 - 1;
+    // ReSharper disable once PossibleLossOfFraction
+    private const double QuantizationStep = (Max - Min) / (double)QuantizationMax;
+    
+    [MethodImpl(MethodImplOptions.AggressiveInlining | MethodImplOptions.AggressiveOptimization)]
+    public static long Encode(long entryId, short count, TermIdMask containerType)
+    {
+        Debug.Assert((count & ~Mask) == 0);
+        return count << ContainerTypeOffset | entryId << EntryIdOffset | (long)containerType;
+    }
+
+    /// <summary>
+    /// Returns id of entry without offset and frequency
+    /// </summary>
+    /// <param name="entryId"></param>
+    /// <returns></returns>
+    public static (long EntryId, short Frequency) Decode(long entryId)
+    {
+        return (entryId >> EntryIdOffset, (short)((entryId >> ContainerTypeOffset) & Mask));
+    }
+
+    /// <summary>
+    /// Container id (page) without encodings.
+    /// </summary>
+    public static long GetContainerId(long entryId)
+    {
+        return (entryId >> EntryIdOffset);
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining | MethodImplOptions.AggressiveOptimization)]
+    public static unsafe void Encode(Span<long> entries, ReadOnlySpan<short> frequencies, bool usePtrVersion = false)
+    {
+        Debug.Assert(frequencies.Length == entries.Length);
+
+        if (usePtrVersion == false)
+            goto Classic;
+        
+        fixed (long* entriesPtr = entries)
+        fixed (short* frequenciesPtr = frequencies)
+        {
```

Comment:
Can we get away without the `fixed` ? I mean using `ref` and `Unsafe` to access without bound checking. Fixing spans to access the pointers is very costly on stuff that has to happen many times repeatedly. It also has the by-product of not needing the `Classic` version.

Reply:
I will leave all optimization to another ticket since this is huge already.

Reply:
https://issues.hibernatingrhinos.com/issue/RavenDB-20044/Optimize-EntryIdEncodings-in-Corax.